sessionctx/sessionstates/session_token.go

```diff
@@ -0,0 +1,337 @@
+//
+// The highlight is that the certificates on all the servers should be the same all the time.
+// However, the certificates should be rotated periodically. Just in case of using different certificates to
+// sign and check, a server should keep the old certificate for a while. A server will try both
+// the 2 certificates to check the signature.
+const (
+	// A token needs a lifetime to avoid brute force attack.
+	tokenLifetime = time.Minute
+	// Reload the certificate periodically because it may be rotated.
+	loadCertInterval = 10 * time.Minute
+	// After a certificate is replaced, it's still valid for oldCertValidTime.
+	// oldCertValidTime must be a little longer than loadCertInterval, because the previous server may
+	// sign with the old cert but the new server checks with the new cert.
+	// - server A loads the old cert at 00:00:00.
+	// - the cert is rotated at 00:00:01 on all servers.
+	// - server B loads the new cert at 00:00:02.
+	// - server A signs token with the old cert at 00:10:00.
+	// - server B reloads the same new cert again at 00:10:01, and it has 3 certs now.
+	// - server B receives the token at 00:10:02, so the old cert should be valid for more than 10m after replacement.
+	oldCertValidTime = 15 * time.Minute
+)
+
+// SessionToken represents the token used to authenticate with the new server.
+type SessionToken struct {
+	Username   string    `json:"username"`
+	SignTime   time.Time `json:"sign-time"`
+	ExpireTime time.Time `json:"expire-time"`
+	Signature  []byte    `json:"signature,omitempty"`
+}
+
+// CreateSessionToken creates a token for the proxy.
+func CreateSessionToken(username string) (*SessionToken, error) {
+	now := getNow()
+	token := &SessionToken{
+		Username:   username,
+		SignTime:   now,
+		ExpireTime: now.Add(tokenLifetime),
+	}
+	tokenBytes, err := json.Marshal(token)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if token.Signature, err = globalSigningCert.sign(tokenBytes); err != nil {
+		return nil, ErrCannotMigrateSession.GenWithStackByArgs(err.Error())
+	}
+	return token, nil
+}
+
+// ValidateSessionToken validates the token sent from the proxy.
+func ValidateSessionToken(tokenBytes []byte, username string) (err error) {
+	var token SessionToken
+	if err = json.Unmarshal(tokenBytes, &token); err != nil {
+		return errors.Trace(err)
+	}
+	signature := token.Signature
+	// Clear the signature and marshal it again to get the original content.
+	token.Signature = nil
+	if tokenBytes, err = json.Marshal(token); err != nil {
+		return errors.Trace(err)
+	}
+	if err = globalSigningCert.checkSignature(tokenBytes, signature); err != nil {
+		return ErrCannotMigrateSession.GenWithStackByArgs(err.Error())
+	}
+	now := getNow()
+	if now.After(token.ExpireTime) {
+		return ErrCannotMigrateSession.GenWithStackByArgs("token expired")
+	}
+	if token.SignTime.Add(tokenLifetime).Before(now) {
```

Comment:
Since `ExpireTime = SignTime + tokenLifeTime`, what is the difference between this and `now.After(token.ExpireTime)`(line 110)?

Reply:
I check both the SignTime and ExpireTime for better safety.
One may forge a very long lifetime to brute force attack:
- The SignTime is now, say, 2022-7-1.
- The token lifetime is set to 1 year, so the ExpireTime is 2023-7-1.
- At a year later, at 2023-6-30, he finally gets the right signature.

Reply:
More specifically, I should check `now.Before(token.ExpireTime) && now.After(token.SignTime) && token.SignTime.Add(tokenLifetime).Equal(token.ExpireTime)`.
However, there may be 2 problems:
- The `tokenLifetime` may change between TiDB versions.
- There may be time bias between TiDB instances.
I want to have a tolerance for these situations.